### ISSUE TITLE
Posting too often, too many false positives

### ISSUE THREAD

**Issue by nineclicks** [0Hughman0/pyredditformatbot#2]
This bot is far too sensitive to false positives and is getting very annoying, some examples:

https://www.reddit.com/r/learnpython/comments/tx5rd4/trouble_getting_the_requested_value_to_be/

https://www.reddit.com/r/learnpython/comments/tx52vp/uppercase_letters_issue_with_python_coding/

**Comment by nineclicks** [comment]
Note that in the first link, the code _was_ formatted properly to begin with and seems to have been set off by `[34,9]`

**Comment by gavin-lb** [comment]
I disagree that the bot is far too sensitive and that it's annoying. I think it is very accurate for the most part and it is actually very nice to not have to tell people to fix their formatting all the time. 

I believe that a lot of the "false positives" you think you're seeing are in fact cases of people ninja editing their posts in response to the bot's comment (on reddit if someone edits within the first few minutes it will not show as having been edited). 

Case in point: it seems like your first example is such an occurrence. Because if you try manually using the validator's pattern yourself on the comment's source you'll see that it does not match:

```
>>> import re
>>> pattern = re.compile(r'''
        ^(?:                        # any of the following is on the left-hand margin (not four spaces in)
            try                     # try block
            |class\s.*?             # class detection
            |def\s.*?\(.*?\)        # function detection
            |for\s.*?\sin\s.*?      # for loop detection
            |while\s.*?             # while loop detection
            |if\s.*?                # if block detection
            |with\s.*?              # with block detection
            |match\s.*?             # match block detection
        ):                          # END NON-CAPTURE GROUP -- literal colon
        ''', re.VERBOSE | re.MULTILINE)
>>> 
>>> pattern.search('''The auto grader on my assignment is looking for the result to look like

    [34,9]

instead I'm getting

    34 9

My code looks like:

    ​numbers = [34, 9, 32, 91, 58, 13, 77, 21, 56] 
    print("Numbers:", numbers)
    n = int(input("Number of elements to fetch from array: "))
    for i in range(n):
    print(numbers[i],end = " ")

&#x200B;''')
>>> 
```

So the comment must have been edited. 

However, in the second case it is in fact a false positive: 

```
>>> pattern.search('''    def frequency_analysis(txt):
        alphabet = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u',
                    'v', 'w', 'x', 'y', 'z']
        file = open(txt,'r')
        text = file.read()
        for letter in alphabet:
            lettercount = re.findall(letter,text)
            print(letter,len(lettercount))
    frequency_analysis("dutch1.txt")

I just wrote this piece of code to count every letter in a txt file. however there is an issue with it:

if i'm not mistaken: my code doesnt count capital letters, which is obviously a problem since they should be accounted for. i know i can convert a string to lowercase with: lower(string) however when i use the command like this:   
text = lower([file.read](https://file.read)())

i get an unresolved reference message, what can i do about that?''')
<re.Match object; span=(527, 547), match="if i'm not mistaken:">
```

But this seems quite unfortunate and I wouldn't imagine such false positives to be a regular occurrence.

**Comment by gavin-lb** [comment]
A possible fix would be to change the pattern to the following:

```
re.compile(r'''
        ^(?:                        # any of the following is on the left-hand margin (not four spaces in)
            try                     # try block
            |class\s.*?             # class detection
            |def\s.*?\(.*?\)        # function detection
            |for\s.*?\sin\s.*?      # for loop detection
            |while\s.*?             # while loop detection
            |if\s.*?                # if block detection
            |with\s.*?              # with block detection
            |match\s.*?             # match block detection
        ):\s*?\n                    # END NON-CAPTURE GROUP -- literal colon, newline
        ''', re.VERBOSE | re.MULTILINE)
```
(where it now checks for the presence of a newline after the colon)

However, whilst this would avoid possible false positive as above, it would allow for the following possible false negatives: 

 - Single line blocks, eg.
```
if var:  return 'Yes'
```

 - Blocks with comments on the first line, eg. 
```
if var:  # comment here
    return 'Yes'
```

Although, the latter issue could be fixed by using the pattern
```
pattern = re.compile(r'''
        ^(?:                        # any of the following is on the left-hand margin (not four spaces in)
            try                     # try block
            |class\s.*?             # class detection
            |def\s.*?\(.*?\)        # function detection
            |for\s.*?\sin\s.*?      # for loop detection
            |while\s.*?             # while loop detection
            |if\s.*?                # if block detection
            |with\s.*?              # with block detection
            |match\s.*?             # match block detection
        ):\s*?(\#.*)?\n             # END NON-CAPTURE GROUP -- literal colon, optional comment, new line
        ''', re.VERBOSE | re.MULTILINE)
```
but this is still susceptible false negatives with single line blocks.

**Comment by 0Hughman0** [comment]
Hi all.

Really appreciate the reports and discussion here regarding how often the bot is posting and the accuracy of its posts.

Particularly @gavin-lb, really appreciate the PR!

You will see that I have tried to re-implement the detection of improperly formatted code by leveraging the `ast` module. This should stop the bot from getting tripped up by sentences that end with a colon! It is now using the Python interpreter to determine if a block of text is code, rather than a super elaborate regex.

I think this should reduce the number of false positives, but also potentially false negatives, because the bot will miss invalid, poorly formatted code.

My plan is to implement a bot summoning feature in order to overcome this.

I hope the more robust testing framework I have set up will aid in this.

I will close this issue for now, and if it starts to cause more problems again, I will happily re-open it.

Best,

Hugh
